src/clients/c++/library/request_grpc.cc

```diff
@@ -1073,26 +1046,41 @@ InferGrpcStreamContextImpl::Run(ResultMap* results)
 {
   // Actually calling AsyncRun() and GetAsyncRunResults()
   std::shared_ptr<Request> req;
-  Error err = AsyncRun(&req);
+  bool callback_invoked = false;
+  std::mutex mtx;
+  std::condition_variable cv;
+  Error err =
+      AsyncRun([&](InferContext* ctx, std::shared_ptr<Request> request) {
+        // Defer the response retrieval to main thread
+        std::lock_guard<std::mutex> lk(mtx);
+        callback_invoked = true;
+        req = std::move(request);
+        cv.notify_all();
+        return;
+      });
   if (!err.IsOk()) {
     return err;
   }
-  bool is_ready;
-  return GetAsyncRunResults(results, &is_ready, req, true);
+  // Wait for the callback
+  {
+    std::unique_lock<std::mutex> lk(mtx);
+    cv.wait(lk, [&callback_invoked]() { return callback_invoked; });
+  }
+  return GetAsyncRunResults(req, results);
 }
 
 Error
 InferGrpcStreamContextImpl::AsyncRun(
     std::shared_ptr<Request>* async_request, OnCompleteFn callback)
 {
   GrpcRequestImpl* current_context;
+  current_context =
+      new GrpcRequestImpl(async_request_id_++, std::move(callback));
+  async_request->reset(static_cast<Request*>(current_context));
+  uintptr_t run_index = current_context->Id();
```

Comment:
Don't use run_index as the object that grpc returns in AsyncTransfer. Just use the shared_ptr<Request> object directly. That should allow you to not need to use ongoing_async_requests_

Reply:
Unfortunately, grpc ClientReaderWriter sync_stream (current service implementation) interface does not allow sending void* pointer to Write and Read interface. ClientAsyncReaderWriter async_stream.h  on the other hand does allow sending and receiving void* tag.

I have removed the ongoing requests from the non-streaming case but not yet for streaming case. The order of the requests and repsonses should be the same for streaming, with this assumption we might be able to replace ongoing_async_requests_ (with a loss in generality) with a lock-free queue storing the callbacks . Shall I do that?

Reply:
I wouldn't since we may be using this same client for the future unordered response case. For now just keep ongoing_async_requests for streaming case.